datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/KafkaPositionTracker.java

```diff
@@ -541,20 +573,79 @@ protected void runOneIteration() {
         }
       }));
 
-      // Query the broker for its offsets for those partitions
       try {
-        _kafkaPositionTracker.queryBrokerForLatestOffsets(_consumer, partitionsNeedingUpdate);
+        if (_enablePartitionLeadershipCalculation) {
+          updatePartitionLeadershipMap();
+        }
       } catch (Exception e) {
-        LOG.warn("Failed to query latest broker offsets via endOffsets() RPC", e);
-        throw e;
+        LOG.warn("Failed to update the partition leadership map. Using stale leadership data. Reason: {}", e.getMessage());
+        LOG.debug("Failed to update the partition leadership map. Using stale leadership data.", e);
+      }
+
+      // Query the broker for its offsets for those partitions
+      batchPartitionsByLeader(partitionsNeedingUpdate).forEach(partitionLeaderBatch -> {
+        try {
+          _kafkaPositionTracker.queryBrokerForLatestOffsets(_consumer, partitionLeaderBatch, _consumerRequestTimeout);
+        } catch (Exception e) {
+          LOG.warn("Failed to query latest broker offsets for this leader batch via endOffsets() RPC. Reason: {}", e.getMessage());
```

Comment:
nit: might want to emit metric here so that you can keep track?

Reply:
This PR is getting long already, and it would scope even more files into this. Can we take this up as a separate item?